File: tools/analysis/field_control.py
"""Reconstructed field direction and jump-request handlers.

Original field functions 0x8009f5a8/0x8009f5f4 and predicate 0x80081f5c.
This models requests, not motion integration. The active encounter-selection
service is unresolved and raises explicitly; its early exits are recovered.
No original tables or RAM addresses form part of the semantic input API.
"""

from __future__ import annotations

from dataclasses import dataclass, replace

from tools.analysis.arithmetic import signed16


class ControlError(ValueError):
    pass


class UnrecoveredEncounter(ControlError):
    pass


def integer(value: int, bits: int, *, signed: bool = False) -> None:
    low = -(1 << (bits - 1)) if signed else 0
    high = (1 << (bits - int(signed))) - 1
    if type(value) is not int or not low <= value <= high:
        raise ControlError(f"control value must fit {'i' if signed else 'u'}{bits}")


@dataclass(frozen=True)
class ControlActor:
    flags: int
    terrain_flags: int
    integer_position: tuple[int, int, int]
    cached_position: tuple[int, int, int]
    direction: int
    animation_mode: int
    pc: int

    def __post_init__(self) -> None:
        integer(self.flags, 32)
        integer(self.terrain_flags, 32)
        for position in (self.integer_position, self.cached_position):
            if len(position) != 3:
                raise ControlError("control position needs three signed coordinates")
            for value in position:
                integer(value, 16, signed=True)
        for value in (self.direction, self.animation_mode, self.pc):
            integer(value, 16)


@dataclass(frozen=True)
class EncounterGates:
    field_active: int
    gate_e4: int
    gate_ec: int
    music_result: int
    encounter_gate: int
    inhibition: int
    menu_gate: int
    enabled_byte: int

    def __post_init__(self) -> None:
        for value in (
            self.field_active,
            self.gate_e4,
            self.gate_ec,
            self.music_result,
            self.encounter_gate,
            self.menu_gate,
        ):
            integer(value, 32)
        integer(self.inhibition, 16, signed=True)
        integer(self.enabled_byte, 8)


def encounter_early_exit(gates: EncounterGates) -> str | None:
    """First return in 0x80079288, before counters, RNG or requests change."""
    for blocked, reason in (
        (gates.field_active == 0, "field_inactive"),
        (gates.gate_e4 == 0, "gate_e4_clear"),
        (gates.gate_ec == 0, "gate_ec_clear"),
        (gates.music_result == 0xFFFFFFFF, "music_pending"),
        (gates.encounter_gate == 0, "encounter_gate_clear"),
        (gates.inhibition == -1, "inhibited"),
        (gates.menu_gate == 1, "menu_gate_one"),
        (gates.enabled_byte == 0, "enabled_byte_clear"),
    ):
        if blocked:
            return reason
    return None


@dataclass(frozen=True)
class ControlInputs:
    held_buttons: int
    pressed_buttons: int
    dialogue_status: tuple[int, int, int, int]
    preserve_nonplayer_motion: int
    jump_contact: int
    jump_mode: int
    repeat_delay: int
    jump_setting: int
    alternate_directions: int
    direction_tables: tuple[tuple[int, ...], tuple[int, ...]]
    camera_angle: int
    encounter: EncounterGates

    def __post_init__(self) -> None:
        for value in (self.held_buttons, self.pressed_buttons, self.repeat_delay):
            integer(value, 16)
        for value in (self.preserve_nonplayer_motion, self.alternate_directions):
            integer(value, 8)
        for value in (self.jump_contact, self.jump_setting):
            integer(value, 32)
        for value in (self.jump_mode, self.camera_angle):
            integer(value, 16, signed=True)
        if len(self.dialogue_status) != 4:
            raise ControlError("control checks exactly four dialogue statuses")
        for value in self.dialogue_status:
            integer(value, 16, signed=True)
        if len(self.direction_tables) != 2 or any(len(t) != 16 for t in self.direction_tables):
            raise ControlError("control needs two original 16-entry direction tables")
        for table in self.direction_tables:
            for value in table:
                integer(value, 16)


@dataclass(frozen=True)
class ControlState:
    stationary_counter: int
    repeat_remaining: int
    latched_jump_setting: int
    updated: int
    break_requested: int

    def __post_init__(self) -> None:
        for value in (self.stationary_counter, self.repeat_remaining):
            integer(value, 16)
        for value in (self.latched_jump_setting, self.updated, self.break_requested):
            integer(value, 32)


@dataclass(frozen=True)
class ControlEffect:
    actor: ControlActor
    state: ControlState
    eligibility: str
    jump: str
    # Ordered nested calls make unobserved/omitted work visible to comparisons.
    calls: tuple[str, ...]
    encounter_exit: str | None
    terrain_call_variant: str | None


def terrain_jump_result(actor: ControlActor) -> int:
    return -int(bool(((actor.flags >> 9) & 3) & (actor.terrain_flags >> 3)))


def request_control(
    opcode: int, actor: ControlActor, state: ControlState, inputs: ControlInputs
) -> ControlEffect:
    """Apply primary a7 or looping wrapper 0c with exact integer widths.

    Blocked requests still advance a7's PC. Wrapper 0c restores that PC and
    requests a scheduler break, including when its inner call is blocked.
    Direction input also invokes the original encounter poll: the recovered
    early-return cases are supported, and active selection is explicit unknown.
    """
    if type(opcode) is not int or opcode not in (0x0C, 0xA7):
        raise ControlError("player control requires primary 0c or a7")
    original_pc = actor.pc
    calls: list[str] = []
    encounter_exit = None
    terrain_call_variant = None
    jump = "control_unavailable"
    if not actor.flags & 0x4000:
        eligibility = "not_control_owner"
        if inputs.preserve_nonplayer_motion == 0:
            actor = replace(actor, flags=actor.flags | 0x01000000)
    elif 0 in inputs.dialogue_status or inputs.encounter.inhibition != 0:
        eligibility = "dialogue" if 0 in inputs.dialogue_status else "inhibited"
        actor = replace(actor, direction=0x8000)
    else:
        eligibility = "available"
        if inputs.held_buttons >> 12:
            calls.append("encounter")
            encounter_exit = encounter_early_exit(inputs.encounter)
            if encounter_exit is None:
                raise UnrecoveredEncounter("active direction-triggered encounter selection")
        counter = state.stationary_counter
        if actor.terrain_flags & 0x00400000:
            if actor.cached_position == actor.integer_position:
                counter = (counter + 1) & 0xFFFF
            # Movement retains the counter on this terrain; it does not reset it.
        else:
            counter = 0
        exceeded = signed16(counter) > 32
        state = replace(state, stationary_counter=32 if exceeded else counter, updated=1)
        held_retry = (
            exceeded
            and bool(inputs.held_buttons & 0x80)
            and not actor.flags & 0x1800
            and inputs.jump_contact == 0xFF
        )
        pressed = bool(inputs.pressed_buttons & 0x80)
        jump = "not_pressed"
        alternate = inputs.jump_mode != 0 and not held_retry
        test_terrain = False
        if held_retry:
            test_terrain = True
        elif alternate:
            if pressed:
                if state.repeat_remaining:
                    jump = "repeat_delay"
                elif inputs.jump_contact != 0xFF:
                    jump = "contact"
                else:
                    test_terrain = True
        elif pressed:
            if actor.flags & 0x1800:
                jump = "jump_or_airborne"
            elif actor.terrain_flags & 0x00400000:
                jump = "counter_terrain"
            elif inputs.jump_contact != 0xFF:
                jump = "contact"
            else:
                test_terrain = True
        if test_terrain:
            calls.append("terrain")
            terrain_call_variant = "alternate" if alternate else "normal"
            if terrain_jump_result(actor):
                jump = "terrain_mask"
            else:
                jump = "held_retry" if held_retry else "pressed"
                actor = replace(
                    actor,
                    flags=actor.flags | 0x800,
                    animation_mode=0xFF if alternate else actor.animation_mode,
                )
                state = replace(state, latched_jump_setting=inputs.jump_setting)
                if alternate:
                    state = replace(state, repeat_remaining=inputs.repeat_delay)
        if alternate and state.repeat_remaining:
            state = replace(state, repeat_remaining=(state.repeat_remaining - 1) & 0xFFFF)
        table = inputs.direction_tables[int(inputs.alternate_directions != 0)]
        direction = table[(inputs.held_buttons >> 12) ^ 15]
        if not direction & 0x8000:
            direction = (direction - inputs.camera_angle) & 0xFFF
        actor = replace(actor, direction=direction)
    actor = replace(actor, pc=(actor.pc + 1) & 0xFFFF)
    if opcode == 0x0C:
        actor = replace(actor, pc=original_pc)
        state = replace(state, break_requested=1)
    return ControlEffect(
        actor, state, eligibility, jump, tuple(calls), encounter_exit, terrain_call_variant
    )

File: src/reconstruction/field_control.cpp
#include "xem/reconstruction/field_control.hpp"

#include <algorithm>
#include <bit>

namespace xem::reconstruction::field {

ControlActor original::read_control_actor(std::span<const std::uint8_t, 0x138> bytes) {
    const auto word = [&](std::size_t offset) {
        return static_cast<std::uint16_t>(static_cast<std::uint16_t>(bytes[offset]) |
                                          (static_cast<std::uint16_t>(bytes[offset + 1]) << 8U));
    };
    ControlActor result;
    result.terrain_flags =
        static_cast<std::uint32_t>(word(0x14)) | (static_cast<std::uint32_t>(word(0x16)) << 16U);
    for (std::size_t coordinate = 0; coordinate < 3; ++coordinate) {
        result.integer_position[coordinate] =
            std::bit_cast<std::int16_t>(word(0x22 + coordinate * 4));
        result.cached_position[coordinate] =
            std::bit_cast<std::int16_t>(word(0x68 + coordinate * 2));
    }
    result.direction = word(0x104);
    result.animation_mode = word(0xe8);
    return result;
}

EncounterExit encounter_early_exit(const BattleRequestState &battle, const EventControl &events,
                                   const EncounterInputs &inputs) noexcept {
    if (battle.field_active == 0)
        return EncounterExit::field_inactive;
    if (events.gate_values[1] == 0)
        return EncounterExit::gate_e4_clear;
    if (events.gate_values[2] == 0)
        return EncounterExit::gate_ec_clear;
    if (inputs.music_result == 0xffffffffU)
        return EncounterExit::music_pending;
    if (inputs.encounter_gate == 0)
        return EncounterExit::encounter_gate_clear;
    if (inputs.inhibition == -1)
        return EncounterExit::inhibited;
    if (battle.menu_gate == 1)
        return EncounterExit::menu_gate_one;
    if (inputs.enabled_byte == 0)
        return EncounterExit::enabled_byte_clear;
    return EncounterExit::none;
}

std::int32_t terrain_jump_result(std::uint32_t actor_flags, std::uint32_t terrain_flags) noexcept {
    return (((actor_flags >> 9U) & 3U) & (terrain_flags >> 3U)) != 0 ? -1 : 0;
}

// EVID-REF-021; corrected field overlay event_source_overlay_sha256.
// Preserve the original branching around held retry and alternate cooldown.
ControlEffect request_field_control(EventContext &context, ControlActor &actor, ControlState &state,
                                    const ControlInputs &inputs, const BattleRequestState &battle) {
    if (context.current_actor == nullptr)
        throw EventError("Control request requires the current event actor");
    auto &event_actor = *context.current_actor;
    ControlEffect effect;
    if ((event_actor.flags & 0x4000U) == 0) {
        if (inputs.preserve_nonplayer_motion == 0)
            event_actor.flags |= 0x01000000U;
    } else if (std::ranges::find(inputs.dialogue_status, 0) != inputs.dialogue_status.end() ||
               inputs.encounter.inhibition != 0) {
        effect.eligibility =
            std::ranges::find(inputs.dialogue_status, 0) != inputs.dialogue_status.end()
                ? ControlEligibility::dialogue
                : ControlEligibility::inhibited;
        actor.direction = 0x8000;
    } else {
        effect.eligibility = ControlEligibility::available;
        if ((inputs.held_buttons >> 12U) != 0) {
            effect.calls[effect.call_count++] = ControlCall::encounter;
            effect.encounter_exit = encounter_early_exit(battle, context.control, inputs.encounter);
            if (effect.encounter_exit == EncounterExit::none)
                throw UnrecoveredEncounter("Active direction-triggered encounter selection");
        }
        state.updated = 1;
        if ((actor.terrain_flags & 0x00400000U) == 0) {
            state.stationary_counter = 0;
        } else if (actor.integer_position == actor.cached_position) {
            state.stationary_counter = static_cast<std::uint16_t>(state.stationary_counter + 1U);
        }
        const bool exceeded = std::bit_cast<std::int16_t>(state.stationary_counter) > 32;
        if (exceeded)
            state.stationary_counter = 32;
        const bool held_retry = exceeded && (inputs.held_buttons & 0x80U) != 0 &&
                                (event_actor.flags & 0x1800U) == 0 && inputs.jump_contact == 0xff;
        const bool pressed = (inputs.pressed_buttons & 0x80U) != 0;
        const bool alternate = inputs.jump_mode != 0 && !held_retry;
        effect.jump = JumpRequest::not_pressed;
        bool test_terrain = false;
        if (held_retry) {
            test_terrain = true;
        } else if (alternate) {
            if (pressed) {
                if (state.repeat_remaining != 0)
                    effect.jump = JumpRequest::repeat_delay;
                else if (inputs.jump_contact != 0xff)
                    effect.jump = JumpRequest::contact;
                else
                    test_terrain = true;
            }
        } else if (pressed) {
            if ((event_actor.flags & 0x1800U) != 0)
                effect.jump = JumpRequest::jump_or_airborne;
            else if ((actor.terrain_flags & 0x00400000U) != 0)
                effect.jump = JumpRequest::counter_terrain;
            else if (inputs.jump_contact != 0xff)
                effect.jump = JumpRequest::contact;
            else
                test_terrain = true;
        }
        if (test_terrain) {
            effect.calls[effect.call_count++] = ControlCall::terrain;
            effect.terrain_call = alternate ? TerrainCall::alternate : TerrainCall::normal;
            if (terrain_jump_result(event_actor.flags, actor.terrain_flags) != 0) {
                effect.jump = JumpRequest::terrain_mask;
            } else {
                effect.jump = held_retry ? JumpRequest::held_retry : JumpRequest::pressed;
                event_actor.flags |= 0x800U;
                state.latched_jump_setting = inputs.jump_setting;
                if (alternate) {
                    actor.animation_mode = 0xff;
                    state.repeat_remaining = inputs.repeat_delay;
                }
            }
        }
        if (alternate && state.repeat_remaining != 0)
            --state.repeat_remaining;
        const auto &table = inputs.direction_tables[inputs.alternate_directions != 0 ? 1 : 0];
        auto direction = table[(inputs.held_buttons >> 12U) ^ 15U];
        if ((direction & 0x8000U) == 0) {
            const auto rotated = static_cast<std::int32_t>(direction) - inputs.camera_angle;
            direction = static_cast<std::uint16_t>(static_cast<std::uint32_t>(rotated) & 0xfffU);
        }
        actor.direction = direction;
    }
    event_actor.pc = static_cast<std::uint16_t>(event_actor.pc + 1U);
    return effect;
}

ControlEffect execute_control_event(EventContext &context, std::uint8_t opcode, ControlActor &actor,
                                    ControlState &state, const ControlInputs &inputs,
                                    const BattleRequestState &battle) {
    if (context.current_actor == nullptr)
        throw EventError("Control event requires the current event actor");
    const auto pc = context.current_actor->pc;
    if (context.program.byte(pc) != opcode)
        throw EventError("Dispatched control opcode disagrees with the working PC");
    if (opcode != 0xa7 && opcode != 0x0c)
        throw UnsupportedInstruction(pc, opcode);
    const auto effect = request_field_control(context, actor, state, inputs, battle);
    if (opcode == 0x0c) {
        context.control.break_requested = 1;
        context.current_actor->pc = pc;
    }
    return effect;
}

} // namespace xem::reconstruction::field
